### DISCUSSION TITLE
Liquid render tag

### DISCUSSION THREAD

**Discussion by renestalder** [11ty/eleventy#1579] [Q&A] [answered]
11ty uses Liquidjs to render Liquid templates. And the Liquidjs documentations says, the [`include`](https://liquidjs.com/tags/include.html) tag is deprecated and that the [`render`](https://liquidjs.com/tags/render.html) tag should be used. Yet when I try to use the render tag in 11ty, I always get the error that this tag doesn't exist. Based on the liquidjs version that 11ty uses, the render tag should be available.

Is there something missing from 11ty or is this a general problem with Liquidjs?

**Comment by pdehaan** [comment, marked as answer]
I think it might depend on how you install 11ty (npm vs GitHub).

I found https://github.com/harttle/liquidjs/issues/163 which mentions that LiquidJS support for the `render` tag was added in [v9.2.0](https://github.com/harttle/liquidjs/releases/tag/v9.2.0).

If I look at the current/latest 11ty package.json (`master` branch), it appears that the LiquidJS version is [`^9.16.1`](https://github.com/11ty/eleventy/blob/904234193766aa9d9a7fb1e5037dea8838995a0f/package.json#L106) (yay!). Although, if I look at the 11ty [`v0.11.x` branch](https://github.com/11ty/eleventy/blob/094c9851d98d77fdd6723673c71cff32de9a7f0a/package.json#L109), it appears that the LiquidJS version is set to `^6.4.3` (not yay), which might explain why [if you're installing via npm] the included version of LiquidJS isn't aware of the `render` tag.

---

Actually, here's a little snippet that I dropped in my local .eleventy.js config file to test:

```js
const eleventy = require("@11ty/eleventy/package.json");
const liquidjs = require("liquidjs/package.json");
console.log(`${eleventy.name} v${eleventy.version} includes ${liquidjs.name}@${liquidjs.version}`);
```

### OUTPUT

> @11ty/eleventy v0.11.1 includes liquidjs@6.4.3

If I reinstall 11ty bleeding edge from GitHub (which I don't really recommend personally) via `npm i github:11ty/eleventy -D`, then rebuild my site, my console now reports:

> @11ty/eleventy v1.0.0-beta.0 includes liquidjs@9.19.0

https://www.11ty.dev/docs/languages/liquid/#optional-set-your-own-library-instance has some instructions for shimming in your own version of LiquidJS if you want, but I don't remember if that exact code snippet works anymore or if some LiquidJS API changes might require you to do a bit of tweaking (see https://github.com/11ty/11ty-website/pull/326 for a rough start).

**Reply by renestalder** [reply]
Thank you @pdehaan. That's probably it. I only checked master, that's where my confusion was coming from. I find it rather weird that the versions on master evolved so much since the last npm publish. Quite a gap between 9.2.0 and 9.16.0, and in fact it seems rather trivial and without much risk to publish minor dependency updates via patches.

I then probably go with initializing my own liquidjs installation. I wanted to do that as a last instance, but that seems the best solution.

Thanks again.

**Reply by pdehaan** [reply]
OK, had a few minutes to mess around with it tonight and installing via the GitHub repo's latest commit (eleventy v1.0.0-beta.0; via <kbd>npm i github:11ty/eleventy -D</kbd>) seems to let me use liquidjs@9 and the `{% render %}` tag without any additional tricks or workarounds.

https://github.com/pdehaan/11ty-liquid9-render

```liquid
{% render header.liquid, name:"peter" %}

<p>Powered by {% calaphon poweredBy.eleventy %}, and {% calaphon poweredBy.liquidjs %}</p>
```

### OUTPUT

```html
<header>
  <p>This is a rendered header</p>
  <p>Author: peter</p>
  <p>Rank: </p>
</header>

<p>Powered by @11ty/eleventy@1.0.0-beta.0, and liquidjs@9.19.0</p>
```

**UPDATE:** I updated the https://github.com/pdehaan/11ty-liquid9-render repo to use a custom instance of a LiquidJS module (just to confirm it's still possible/easy w/ liquidjs@9). It was a bit hard to prove it was working, but I set the `dynamicPartials` config option to `true` which caused my build to fail since now it expects the paths in `include`/`render` tags to be quoted, but after changing my code to `{% render "header.liquid", name:"peter" %}` it worked fine!

TL;DR: if you install `@11ty/eleventy` via GitHub using <kbd>npm i github:11ty/eleventy</kbd>, it should use the latest version of `liquidjs@9` and everything should work as expected. If you want to use a slightly different version of LiquidJS that is newer than @11ty/eleventy's `"liquidjs": "^9.16.1",` (or just want to hack around w/ liquidjs configs), you can still set a custom LiquidJS library by adding this into your .eleventy.js config file:

```js
module.exports = (eleventyConfig) => {
  const { Liquid } = require("liquidjs");
  eleventyConfig.setLibrary(
    "liquid",
    new Liquid({
      dynamicPartials: true,
      strictFilters: true,
    })
  );

  // ...
  return {
    // ...
  };
};
```
